### PR TITLE
feat: data residency

### PR DESCRIPTION
**Pull request by MoumitaM**
## Description of the change

> This feature will enable the customer to select the dataplane URL based on the region. Dataplane URL for different regions will be provided in the source config response and based on the input provided by the user through loadOptions dataplane Url will be selected.
- We are not deprecating the existing way of providing the dataplane Url. Also that will be used as fallback URL if default region URL is not present in source config response.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/718)

### CODE REVIEW

File: analytics.js [hunk 1]
```diff
@@ -182,6 +185,50 @@ class Analytics {
...
+    // Check if dataPlaneUrls object is present in source config
+    if (Object.keys(dataPlaneUrls).length) {
+      const inputRegion = this.getResidencyServerInput() || '';
+      const dataPlaneUrl =
+        dataPlaneUrls[inputRegion.toUpperCase()] || dataPlaneUrls[inputRegion.toLowerCase()];
+      if (dataPlaneUrl) {
+        return dataPlaneUrl;
+      }
+      if (dataPlaneUrls[DEFAULT_REGION]) {
+        return dataPlaneUrls[DEFAULT_REGION];
```
Comment (saikumarrs) [inline analytics.js:225]: We should check if the user provided URLs are also valid.

File: analytics.js [hunk 2]
```diff
@@ -232,6 +279,20 @@ class Analytics {
         }
       }
 
+      // determine the dataPlaneUrl
+      try {
+        this.serverUrl = this.resolveDataPlaneUrl(response.source.dataPlaneUrls);
+      } catch (e) {
+        handleError(e);
+        this.serverUrl = this.serverUrl ? this.serverUrl : DEFAULT_DATAPLANE_URL;
+      }
```
Comment (saikumarrs) [inline analytics.js:288]: We can move the error handling to the function and simplify the logic better. Upon exceptions or invalid input, you can use the default data-plane URL.

File: __tests__/script.test.js [hunk 3]
```diff
@@ -8,13 +8,20 @@ describe('Test suite for the SDK', () => {
...
+    responseText: JSON.stringify({
+      source: {
+        config: {},
+        destinations: [],
+      },
+    }),
     status: 200,
   };
 
+  xhrMock['send'] = jest.fn(() => xhrMock.onload());
```
Comment (saikumarrs) [inline __tests__/script.test.js:23]: Why is this needed?

File: __tests__/script.test.js [hunk 4]
```diff
@@ -8,13 +8,20 @@ describe('Test suite for the SDK', () => {
   const xhrMock = {
```
Comment (saikumarrs) [inline __tests__/script.test.js:8]: Add unit tests (not integration tests) for all the newly added code.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: analytics.js [hunk 5]
```diff
@@ -860,7 +936,9 @@ class Analytics {
   reset(flag) {
     leaveBreadcrumb(`reset API :: flag: ${flag}`);
 
-    if (!this.loaded) return;
+    if (!this.loaded) {
+      this.toBeProcessedArray.push(['reset', flag]);
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js:940]: If we return the same thing will happen i.e. it will be ignored.

File: dist/rudder-sdk-js/index.d.ts [hunk 6]
```diff
@@ -92,7 +100,8 @@ declare module "rudder-sdk-js" {
     sessions?: {
       autoTrack?: boolean; // Defaults to true
       timeout?: number; // Defaults to 30 mins
-    }
+    };
+    residencyServer: RESIDENCY_SERVER.US | RESIDENCY_SERVER.EU;
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply dist/rudder-sdk-js/index.d.ts:104]: I don't think this will work.

File: analytics.js (hunk 1 above)
Comment (MoumitaM) [reply analytics.js:225]: Do you mean URLs provided in source config?

File: __tests__/script.test.js (hunk 3 above)
Comment (MoumitaM) [reply __tests__/script.test.js:23]: Our source config fetching call uses the `xhr.onload` approach. Previously we didn't bother to check this because xhr.send getting triggered was enough. But now we have to capture source config call and other API calls.

File: analytics.js [hunk 7]
```diff
@@ -951,10 +1033,14 @@ class Analytics {
     }
     if (options && options.cookieConsentManager)
       this.cookieConsentOptions = cloneDeep(options.cookieConsentManager);
-    if (!this.isValidWriteKey(writeKey) || !this.isValidServerUrl(serverUrl)) {
-      throw Error('Unable to load the SDK due to invalid writeKey or serverUrl');
+    if (!this.isValidWriteKey(writeKey)) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 8]
```diff
@@ -458,7 +519,10 @@ class Analytics {
    */
   page(category, name, properties, options, callback) {
     leaveBreadcrumb(`Page event`);
-    if (!this.loaded) return;
+    if (!this.loaded) {
+      this.toBeProcessedArray.push(['page', category, name, properties, options, callback]);
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 9]
```diff
@@ -232,6 +279,20 @@ class Analytics {
         }
       }
 
+      // determine the dataPlaneUrl
+      try {
+        this.serverUrl = this.resolveDataPlaneUrl(response.source.dataPlaneUrls);
+      } catch (e) {
+        handleError(e);
+        this.serverUrl = this.serverUrl ? this.serverUrl : DEFAULT_DATAPLANE_URL;
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 10]
```diff
@@ -182,6 +185,50 @@ class Analytics {
...
+   */
+  checkUrlFromCode() {
+    if (this.serverUrl && this.isValidServerUrl(this.serverUrl)) {
+      return this.serverUrl;
+    }
+    return DEFAULT_DATAPLANE_URL;
+  }
+
+  /**
+   * A function to determine the dataPlaneUrl for sending cloud mode calls
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 11]
```diff
@@ -182,6 +185,50 @@ class Analytics {
     }
   }
 
+  /**
+   * A function to validate and return Residency server input
+   * @returns string/undefined
+   */
+  getResidencyServerInput() {
+    const region = get(this.options, 'residencyServer');
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: dist/rudder-sdk-js/index.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply dist/rudder-sdk-js/index.d.ts]: Fixed another thing.

File: analytics.js [hunk 12]
```diff
@@ -182,6 +185,50 @@ class Analytics {
...
+    return typeof region === 'string' && RESIDENCY_SERVERS.includes(region.toUpperCase())
+      ? region
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 13]
```diff
@@ -182,6 +185,50 @@ class Analytics {
     }
   }
 
+  /**
+   * A function to validate and return Residency server input
+   * @returns string/undefined
+   */
+  getResidencyServerInput() {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 14]
```diff
@@ -182,6 +185,50 @@ class Analytics {
...
+   * A function to determine the dataPlaneUrl for sending cloud mode calls
+   * @param {Object} dataPlaneUrls An object containing dataPlaneUrl for different region
+   * @returns string
+   */
+  resolveDataPlaneUrl(dataPlaneUrls = {}) {
+    // Check if dataPlaneUrls object is present in source config
+    if (Object.keys(dataPlaneUrls).length) {
+      const inputRegion = this.getResidencyServerInput() || '';
+      const dataPlaneUrl =
+        dataPlaneUrls[inputRegion.toUpperCase()] || dataPlaneUrls[inputRegion.toLowerCase()];
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

File: utils/constants.js [hunk 15]
```diff
@@ -33,6 +33,11 @@ const DEFAULT_SESSION_TIMEOUT = 30 * 60 * 1000; // 30 min in milliseconds
 const MIN_SESSION_TIMEOUT = 10 * 1000; // 10 sec in milliseconds
 const MIN_SESSION_ID_LENGTH = 10;
 
+const DEFAULT_REGION = 'US';
+const DEFAULT_DATAPLANE_URL = 'https://hosted.rudderlabs.com';
+
+const RESIDENCY_SERVERS = ['US', 'EU'];
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply utils/constants.js]: Addressed

File: __tests__/script.test.js (hunk 4 above)
Comment (MoumitaM) [reply __tests__/script.test.js:8]: Existing test cases are covering most of the things.

File: analytics.js (hunk 2 above)
Comment (MoumitaM) [reply analytics.js]: This way we can also handle cases if say accessing `response.source.dataPlanes` generates some error.

File: analytics.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply analytics.js:964]: I mean, after you buffer the request.

File: analytics.js (hunk 1 above)
Comment (saikumarrs) [reply analytics.js]: Yes.

File: __tests__/script.test.js (hunk 4 above)
Comment (saikumarrs) [reply __tests__/script.test.js:8]: Not enough. A lot of scenarios are not taken care of in them. Those are e2e tests.

File: analytics.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js:966]: Right I missed that

File: analytics.js (hunk 1 above)
Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js:966]: Addressed

File: dist/rudder-sdk-js/index.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply dist/rudder-sdk-js/index.d.ts]: Addressed

File: __tests__/script.test.js (hunk 4 above)
Comment (MoumitaM) [reply __tests__/script.test.js:8]: Addressed

File: analytics.js [hunk 16]
```diff
@@ -280,12 +235,8 @@ class Analytics {
       }
 
       // determine the dataPlaneUrl
-      try {
-        this.serverUrl = this.resolveDataPlaneUrl(response.source.dataPlaneUrls);
-      } catch (e) {
-        handleError(e);
-        this.serverUrl = this.serverUrl ? this.serverUrl : DEFAULT_DATAPLANE_URL;
-      }
+      this.serverUrl = resolveDataPlaneUrl(response);
```
Comment (saikumarrs) [inline analytics.js:238]: `serverUrl` is not sent here.

Comment (saikumarrs) [reply analytics.js:238]: Also, `options`.

File: utils/utils.js [hunk 17]
```diff
@@ -743,6 +749,77 @@ const getStringId = (id) => {
...
+        const defaultUrl = getDefaultUrlofRegion(regionUrlArr);
+        if (defaultUrl) {
+          return defaultUrl;
+        }
+      }
+    }
+    // return the dataPlaneUrl provided in load API(if available)
+    if (serverUrl && isValidServerUrl(serverUrl)) {
+      return serverUrl;
+    }
```
Comment (saikumarrs) [inline utils/utils.js:814]: The URL provided in load options takes the least preference? Doesn't seem right.

File: utils/utils.js [hunk 18]
```diff
@@ -743,6 +749,77 @@ const getStringId = (id) => {
...
+/**
+ * A function to get url from source config response
+ * @param {array} urls    An array of objects containing urls
+ * @returns
+ */
+const getDefaultUrlofRegion = (urls) => {
+  let url;
+  if (Array.isArray(urls) && urls.length) {
+    const obj = urls.find((elem) => elem.default === true);
+    if (obj && obj.url && isValidServerUrl(obj.url)) {
```
Comment (saikumarrs) [inline utils/utils.js:784]: Redundant.
```suggestion
    if (isValidServerUrl(obj.url)) {
```

File: utils/utils.js [hunk 19]
```diff
@@ -743,6 +749,77 @@ const getStringId = (id) => {
...
+    }
+    // return the dataPlaneUrl provided in load API(if available)
+    if (serverUrl && isValidServerUrl(serverUrl)) {
+      return serverUrl;
+    }
+    // return the default dataPlaneUrl
+    return DEFAULT_DATAPLANE_URL;
+  } catch (e) {
+    handleError(e);
+    return serverUrl || DEFAULT_DATAPLANE_URL;
```
Comment (saikumarrs) [inline utils/utils.js:819]: Don't you have to validate the `serverUrl` here?

File: utils/utils.js [hunk 20]
```diff
@@ -743,6 +749,77 @@ const getStringId = (id) => {
...
+};
+
+/**
+ * A function to determine the dataPlaneUrl
+ * @param {Object} dataPlaneUrls An object containing dataPlaneUrl for different region
+ * @returns string
+ */
+const resolveDataPlaneUrl = (response, serverUrl, options) => {
+  try {
+    const dataPlaneUrls = response.source.dataplanes || {};
```
Comment (saikumarrs) [inline utils/utils.js:798]: Better to not name the variable `Ursl` if it is indeed not.

File: __tests__/data-residency.test.js [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+      US: [
+        {
+          url: 'https://sample.rudderlabs.com/us',
+          default: true,
+        },
+      ],
+      EU: [
+        {
+          url: 'https://sample.rudderlabs.com/eu',
+          default: true,
```
Comment (saikumarrs) [inline __tests__/data-residency.test.js:16]: How can there be two defaults?

File: __tests__/data-residency.test.js [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+        },
+      ],
+    },
+  },
+};
+
+const serverUrl = 'https://sample.rudderlabs.com';
+
+test('When no sc url and no load call url default will be choosen', () => {
```
Comment (saikumarrs) [inline __tests__/data-residency.test.js:51]: `sc url`?

Comment (saikumarrs) [reply __tests__/data-residency.test.js:51]: Be clear.

File: __tests__/data-residency.test.js [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+  const url = resolveDataPlaneUrl({ source: {} }, undefined, undefined);
```
Comment (saikumarrs) [inline __tests__/data-residency.test.js:52]: You don't need to send `undefined` for arguments on the right-hand side to skip them.

File: __tests__/data-residency.test.js [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  expect(url).toEqual(serverUrl);
+});
+
+test('When sc url is provided that will be choosen irrespective of load call url', () => {
+  const url = resolveDataPlaneUrl(sourceConfigWithBothDataPlane, serverUrl, undefined);
+  const defaultRegion = sourceConfigWithBothDataPlane.source.dataplanes['US'][0];
+  expect(url).toEqual(defaultRegion.url);
+});
+
+test('When User provides invalid residency server input + no sc url + no load call url: default will be choosen', () => {
```
Comment (saikumarrs) [inline __tests__/data-residency.test.js:67]: We are testing a function here. No context of `User` is available.

File: __tests__/data-residency.test.js [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+test('When User provides invalid residency server input + no sc url + load call url provided : load call url will be choosen', () => {
+  const url = resolveDataPlaneUrl({ source: {} }, serverUrl, { residencyServer: 'test' });
+  expect(url).toEqual(serverUrl);
+});
+
+test('When User provides invalid residency server input + sc url + load call url provided : sc url will be choosen', () => {
+  const url = resolveDataPlaneUrl(sourceConfigWithBothDataPlane, serverUrl, {
+    residencyServer: 'test',
+  });
+  const defaultRegion = sourceConfigWithBothDataPlane.source.dataplanes['US'][0];
```
Comment (saikumarrs) [inline __tests__/data-residency.test.js:81]: Input and expected values should be part of the test case data itself. Ideally, there should be no logic in test cases.

File: analytics.js (hunk 16 above)
Comment (MoumitaM) [reply analytics.js]: Thank you for pointing out, addressed it

File: utils/utils.js (hunk 18 above)
Comment (MoumitaM) [reply utils/utils.js]: obj might be undefined, in that case, `obj` presence checking is needed omitted `obj.url` checking

File: utils/utils.js (hunk 19 above)
Comment (MoumitaM) [reply utils/utils.js]: Right, Addressed

File: utils/utils.js (hunk 20 above)
Comment (MoumitaM) [reply utils/utils.js]: Addressed

File: __tests__/data-residency.test.js (hunk 21 above)
Comment (MoumitaM) [reply __tests__/data-residency.test.js:19]: Yes there can be. Here is the PRD link, check the source config response.

File: __tests__/data-residency.test.js (hunk 22 above)
Comment (MoumitaM) [reply __tests__/data-residency.test.js]: Addressed

File: __tests__/data-residency.test.js (hunk 23 above)
Comment (MoumitaM) [reply __tests__/data-residency.test.js]: Addressed

File: __tests__/data-residency.test.js (hunk 24 above)
Comment (MoumitaM) [reply __tests__/data-residency.test.js]: Addressed

File: __tests__/data-residency.test.js (hunk 25 above)
Comment (MoumitaM) [reply __tests__/data-residency.test.js]: Addressed

File: utils/utils.js (hunk 17 above)
Comment (MoumitaM) [reply utils/utils.js]: Here is the precedence list:
source config URL > load call URL > default URL

File: __tests__/data-residency.test.js (hunk 25 above)
Comment (saikumarrs) [reply __tests__/data-residency.test.js]: No, that's not what I meant.

```
const testCaseData = [
{
  "description": "A meaningful description 1",
  "input": {
  ...
  },
  "output": {
  ...
  }
},
{
  "description": "A meaningful description 2",
  "input": {
  ...
  },
  "output": {
  ...
  }
},
...
]
```

**Review by saikumarrs — DISMISSED** [review]

**Review by saikumarrs — APPROVED** [review]

File: src/utils/utils.js [hunk 26]
```diff
@@ -679,6 +685,77 @@ const getStringId = (id) => {
...
+          return defaultUrl;
+        }
+      }
+    }
+    // return the dataPlaneUrl provided in load API(if available)
+    if (isValidServerUrl(serverUrl)) {
+      return serverUrl;
+    }
+    // return the default dataPlaneUrl
+    return DEFAULT_DATAPLANE_URL;
```
Comment (bardisg) [inline src/utils/utils.js:752]: Lets change this to trow an error as it used to throw when data plane url was undefined or invalid.
Seems like product is yet to determine how this should behave so ets not introduce the auto fallback to hosted url yet.

Other sdks can do it since they do not autoupdate so breaking changes are esier to apply there.

will create a separate task only for these case

File: src/core/analytics.js [hunk 27]
```diff
@@ -464,7 +476,10 @@ class Analytics {
    */
   page(category, name, properties, options, callback) {
     leaveBreadcrumb(`Page event`);
-    if (!this.loaded) return;
+    if (!this.loaded) {
+      this.toBeProcessedArray.push(['page', ...arguments]);
```
Comment (bardisg) [inline src/core/analytics.js:480]: can you please remind me why we did this one?

Comment (bardisg) [reply src/core/analytics.js:480]: is this tonot loose events if the sdk has loaded but still has not proceesed the config?

File: src/utils/utils.js (hunk 26 above)
Comment (bardisg) [reply src/utils/utils.js:752]: throw Error('Unable to load the SDK due to invalid data plane url');

File: src/core/analytics.js (hunk 27 above)
Comment (MoumitaM) [reply src/core/analytics.js:480]: To buffer the events that are triggered before SDK is loaded.

File: analytics.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: This is an old line. To buffer the events that are triggered before SDK is loaded.

File: analytics.js [hunk 28]
```diff
@@ -182,6 +185,50 @@ class Analytics {
...
+    }
+    return DEFAULT_DATAPLANE_URL;
+  }
+
+  /**
+   * A function to determine the dataPlaneUrl for sending cloud mode calls
+   * @param {Object} dataPlaneUrls An object containing dataPlaneUrl for different region
+   * @returns string
+   */
+  resolveDataPlaneUrl(dataPlaneUrls = {}) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Yes, this is the logic now.

[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: You are looking into some old code.

File: src/utils/utils.js (hunk 26 above)
Comment (MoumitaM) [reply src/utils/utils.js]: Addressed

**Review by bardisg — APPROVED** [review]
